app/src/main/java/com/rkm/tasky/network/repository/implementation/TaskyRemoteRepositoryImpl.kt

```diff
@@ -0,0 +1,164 @@
+package com.rkm.tasky.network.repository.implementation
+
+import com.rkm.tasky.di.IoDispatcher
+import com.rkm.tasky.network.datasource.TaskyRemoteDataSource
+import com.rkm.tasky.network.model.Reminder
+import com.rkm.tasky.network.model.Task
+import com.rkm.tasky.network.model.request.AccessTokenRequest
+import com.rkm.tasky.network.model.request.CreateEventRequest
+import com.rkm.tasky.network.model.request.LoginUserRequest
+import com.rkm.tasky.network.model.request.RegisterUserRequest
+import com.rkm.tasky.network.model.request.SyncAgendaRequest
+import com.rkm.tasky.network.model.request.UpdateEventRequest
+import com.rkm.tasky.network.model.response.AccessTokenResponse
+import com.rkm.tasky.network.model.response.AgendaResponse
+import com.rkm.tasky.network.model.response.AttendeeResponse
+import com.rkm.tasky.network.model.response.EventResponse
+import com.rkm.tasky.network.model.response.LoginUserResponse
+import com.rkm.tasky.network.repository.abstraction.TaskyRemoteRepository
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import retrofit2.HttpException
+import java.io.IOException
+import javax.inject.Inject
+
+class TaskyRemoteRepositoryImpl @Inject constructor(
+    private val dataSource: TaskyRemoteDataSource,
+    @IoDispatcher private val dispatcher: CoroutineDispatcher
+): TaskyRemoteRepository {
+    override suspend fun registerUser(request: RegisterUserRequest): Result<Unit> {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun loginUser(
+        request: LoginUserRequest
+    ): Result<LoginUserResponse> = withContext(dispatcher) {
+        return@withContext try {
+            val response = dataSource.loginUser(request)
+            if (response.isSuccessful && response.body() != null) {
+                Result.success(response.body()!!)
+            } else if (response.body() == null) {
+                Result.failure(NullPointerException())
+            } else {
+                Result.failure(HttpException(response))
+            }
+        } catch (e: IOException) {
+            Result.failure(e)
+        }
```

Comment:
Since you'll have a lot of these API calls that can all give you similar errors, could it make sense extracting this when expression into a function? Something like this maybe?


```
fun <E: Error> Exception.asResult(): Result.Error<E> {
    return when(this) {
        ...
    }
}
``